### ISSUE TITLE
About ST from waymo to nus

### ISSUE THREAD

**Issue by xero0519** [CVMI-Lab/ST3D#32]
hi, Jihan
I have try a lot of times using your lastest config files, but fail to reproduce the performance of ST from waymo to nus.
**for secondiou_st, the best performance showed in No. 4 epoch.**
[2021-12-02 09:53:08,270  eval_utils.py 50  INFO]  *************** EPOCH 4 EVALUATION *****************
[2021-12-02 09:54:26,908  eval_utils.py 89  INFO]  *************** Performance of EPOCH 4 *****************
[2021-12-02 09:54:26,908  eval_utils.py 91  INFO]  Generate label finished(sec_per_example: 0.0131 second).
[2021-12-02 09:54:26,908  eval_utils.py 107  INFO]  recall_roi_0.3: 0.582705
[2021-12-02 09:54:26,908  eval_utils.py 108  INFO]  recall_rcnn_0.3: 0.582705
[2021-12-02 09:54:26,909  eval_utils.py 107  INFO]  recall_roi_0.5: 0.476606
[2021-12-02 09:54:26,909  eval_utils.py 108  INFO]  recall_rcnn_0.5: 0.476606
[2021-12-02 09:54:26,909  eval_utils.py 107  INFO]  recall_roi_0.7: 0.234905
[2021-12-02 09:54:26,909  eval_utils.py 108  INFO]  recall_rcnn_0.7: 0.234905
[2021-12-02 09:54:26,911  eval_utils.py 116  INFO]  Average predicted number of objects(6019 samples): 5.036
[2021-12-02 09:54:37,770  eval_utils.py 127  INFO]  Car AP@0.70, 0.70, 0.70:
bbox AP:45.3321, 45.3321, 45.3321
bev  AP:30.9359, 30.9359, 30.9359
3d   AP:17.2579, 17.2579, 17.2579
aos  AP:26.60, 26.60, 26.60
Car AP_R40@0.70, 0.70, 0.70:
bbox AP:47.2941, 47.2941, 47.2941
bev  AP:26.0944, 26.0944, 26.0944
3d   AP:12.9529, 12.9529, 12.9529
aos  AP:27.46, 27.46, 27.46
Car AP@0.70, 0.50, 0.50:
bbox AP:45.3321, 45.3321, 45.3321
bev  AP:34.6136, 34.6136, 34.6136
3d   AP:33.8733, 33.8733, 33.8733
aos  AP:26.60, 26.60, 26.60
Car AP_R40@0.70, 0.50, 0.50:
bbox AP:47.2941, 47.2941, 47.2941
bev  AP:34.4677, 34.4677, 34.4677
3d   AP:29.9190, 29.9190, 29.9190
aos  AP:27.46, 27.46, 27.46
I used 4 RTX2080TI cards, with batch_size 16. 
Could you help me to find what is wrong with it?

**Comment by jihanyang** [comment]
Sorry, the given information is not enough to have a analysis.

**Comment by xero0519** [comment]
Thanks for your reply. We adopted the waymo-pretrained ROS model that was applied from you and use the lastest da-waymo-to_nus config files 'secondiou_st3d.yaml' to reproduce your experiment. We use tha standard nusence dataset with 28,130 training frames and 6,019 validation  frames and without merge neighbor sweep. There is still a 3%-%4 gap comparing with your released. We run the codes on 4 RTX2080TI cards, with batch_size 16. Hope you can help us to figure out what is wrong.

**Comment by jihanyang** [comment]
May I know the performance of your waymo-pretrained ROS model?

**Comment by xero0519** [comment]
[2021-12-26 10:58:21,119  config.py 12  INFO]  cfg.OPTIMIZATION.LR_WARMUP: False
[2021-12-26 10:58:21,119  config.py 12  INFO]  cfg.OPTIMIZATION.WARMUP_EPOCH: 1
[2021-12-26 10:58:21,119  config.py 12  INFO]  cfg.OPTIMIZATION.GRAD_NORM_CLIP: 10
[2021-12-26 10:58:21,119  config.py 9  INFO]
cfg.SELF_TRAIN = edict()
[2021-12-26 10:58:21,120  config.py 9  INFO]
cfg.SELF_TRAIN.SRC = edict()
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.SRC.USE_DATA: False
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.SRC.USE_GRAD: False
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.SRC.LOSS_WEIGHT: 1.0
[2021-12-26 10:58:21,120  config.py 9  INFO]
cfg.SELF_TRAIN.TAR = edict()
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.TAR.USE_DATA: True
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.TAR.LOSS_WEIGHT: 1.0
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.SCORE_THRESH: [0.5]
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.NEG_THRESH: [0.2]
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.UPDATE_PSEUDO_LABEL: [0]
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.UPDATE_PSEUDO_LABEL_INTERVAL: 1000
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.INIT_PS: None
[2021-12-26 10:58:21,120  config.py 9  INFO]
cfg.SELF_TRAIN.PROG_AUG = edict()
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.PROG_AUG.ENABLED: False
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.PROG_AUG.UPDATE_AUG: [5, 10, 20, 25]
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.PROG_AUG.SCALE: 1.1
[2021-12-26 10:58:21,120  config.py 9  INFO]
cfg.SELF_TRAIN.MEMORY_ENSEMBLE = edict()
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.ENABLED: False
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.NAME: consistency_ensemble
[2021-12-26 10:58:21,120  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.IOU_THRESH: 0.1
[2021-12-26 10:58:21,121  config.py 9  INFO]
cfg.SELF_TRAIN.MEMORY_ENSEMBLE.NMS_CONFIG = edict()
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.NMS_CONFIG.NMS_TYPE: nms_gpu
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.NMS_CONFIG.MULTI_CLASSES_NMS: False
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.NMS_CONFIG.NMS_PRE_MAXSIZE: 512
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.NMS_CONFIG.NMS_POST_MAXSIZE: 100
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.NMS_CONFIG.NMS_THRESH: 0.1
[2021-12-26 10:58:21,121  config.py 9  INFO]
cfg.SELF_TRAIN.MEMORY_ENSEMBLE.MEMORY_VOTING = edict()
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.MEMORY_VOTING.ENABLED: False
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.MEMORY_VOTING.IGNORE_THRESH: 2
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.SELF_TRAIN.MEMORY_ENSEMBLE.MEMORY_VOTING.RM_THRESH: 3
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.TAG: secondiou_st3d
[2021-12-26 10:58:21,121  config.py 12  INFO]  cfg.EXP_GROUP_PATH: da-waymo-nus_models/secondiou_st3d
[2021-12-26 10:58:21,215  nuscenes_dataset.py 25  INFO]  Loading NuScenes dataset
[2021-12-26 10:58:21,761  nuscenes_dataset.py 37  INFO]  Total samples for NuScenes dataset: 6019
[2021-12-26 10:58:24,698  detector3d_template.py 286  INFO]  ==> Loading parameters from checkpoint ../../ST3D_SPC10/pretrained/da-waymo-kitti/ros/secondiou/waymo2kitti_ros_secondiou.pth to CPU
[2021-12-26 10:58:24,861  detector3d_template.py 292  INFO]  ==> Checkpoint trained from version: pcdet+0.2.0+4d3e8b6+py964f663
[2021-12-26 10:58:25,094  detector3d_template.py 311  INFO]  ==> Done (loaded 189/189)
[2021-12-26 10:58:25,114  eval_utils.py 52  INFO]  *************** EPOCH 2 EVALUATION *****************
[2021-12-26 10:59:12,563  eval_utils.py 91  INFO]  *************** Performance of EPOCH 2 *****************
[2021-12-26 10:59:12,563  eval_utils.py 93  INFO]  Generate label finished(sec_per_example: 0.0078 second).
[2021-12-26 10:59:12,563  eval_utils.py 109  INFO]  recall_roi_0.3: 0.599659
[2021-12-26 10:59:12,563  eval_utils.py 110  INFO]  recall_rcnn_0.3: 0.599659
[2021-12-26 10:59:12,563  eval_utils.py 109  INFO]  recall_roi_0.5: 0.476021
[2021-12-26 10:59:12,563  eval_utils.py 110  INFO]  recall_rcnn_0.5: 0.476021
[2021-12-26 10:59:12,563  eval_utils.py 109  INFO]  recall_roi_0.7: 0.214413
[2021-12-26 10:59:12,563  eval_utils.py 110  INFO]  recall_rcnn_0.7: 0.214413
[2021-12-26 10:59:12,566  eval_utils.py 118  INFO]  Average predicted number of objects(6019 samples): 13.397
[2021-12-26 10:59:35,611  eval_utils.py 129  INFO]  Car AP@0.70, 0.70, 0.70:
bbox AP:77.8167, 77.8167, 77.8167
bev  AP:26.7040, 26.7040, 26.7040
3d   AP:15.1878, 15.1878, 15.1878
aos  AP:46.70, 46.70, 46.70
Car AP_R40@0.70, 0.70, 0.70:
bbox AP:81.9946, 81.9946, 81.9946
bev  AP:22.5478, 22.5478, 22.5478
3d   AP:9.8990, 9.8990, 9.8990
aos  AP:45.78, 45.78, 45.78
Car AP@0.70, 0.50, 0.50:
bbox AP:77.8167, 77.8167, 77.8167
bev  AP:36.1997, 36.1997, 36.1997
3d   AP:31.2411, 31.2411, 31.2411
aos  AP:46.70, 46.70, 46.70
Car AP_R40@0.70, 0.50, 0.50:
bbox AP:81.9946, 81.9946, 81.9946
bev  AP:33.3590, 33.3590, 33.3590
3d   AP:27.8346, 27.8346, 27.8346
aos  AP:45.78, 45.78, 45.78

[2021-12-26 10:59:35,634  eval_utils.py 135  INFO]  Result is save to /home/chi_zhang/3d-detection/ST3D_TO_NUS/output/da-waymo-nus_models/secondiou_st3d/secondiou_st3d/default/eval/epoch_2/val/default
[2021-12-26 10:59:35,634  eval_utils.py 136  INFO]  ****************Evaluation done.*****************

**Comment by xero0519** [comment]
hi Jihan, above info includes config details for training.  The test script is  "bash scripts/dist_test.sh 4 --cfg_file cfgs/da-waymo-nus_models/secondiou_st3d/secondiou_st3d.yaml --batch_size 32 --ckpt ../../ST3D_SPC10/pretrained/da-waymo-kitti/ros/secondiou/waymo2kitti_ros_secondiou.pth" hope it would help your analysis.

**Comment by jihanyang** [comment]
Hello, it seems that the relative improvement from pretrained model to the model after self-training behaves normally. And I think the main problem takes place in the pretraining stage (refer to my paper). Another evidence is that the adaptation result from waymo->nus typically very stable across different epochs, so I always just select the last epoch as pretrained model, while you use the 2nd epoch model. 

1. Maybe you can test the last epoch result?
2. You can also request the pretrained model of waymo -> nus (for further verify the self-training stage is normal) by sending email to me with your name, institute, a screenshot of the the Waymo dataset registration confirmation mail, and your intended usage. Please note that Waymo open dataset is under strict non-commercial license so we are not allowed to share the model with you if it will used for any profit-oriented activities.

**Comment by xero0519** [comment]
Thanks. Yes, it is the problem of pre-trained model.
